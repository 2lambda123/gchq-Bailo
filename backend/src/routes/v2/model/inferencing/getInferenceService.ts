import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { AuditInfo } from '../../../../connectors/v2/audit/Base.js'
import audit from '../../../../connectors/v2/audit/index.js'
import { InferenceInterface } from '../../../../models/v2/Inference.js'
import { getInferenceByImage } from '../../../../services/v2/inference.js'
import { inferenceInterfaceSchema, registerPath } from '../../../../services/v2/specification.js'
import { parse } from '../../../../utils/v2/validate.js'

export const getInferenceSchema = z.object({
  params: z.object({
    modelId: z.string(),
    image: z.string(),
    tag: z.string(),
  }),
})

registerPath({
  method: 'get',
  path: '/api/v2/model/{modelId}/inference/{image}/{tag}',
  tags: ['inference'],
  description: 'Get details for an inferencing service within the cluster.',
  schema: getInferenceSchema,
  responses: {
    200: {
      description: 'Details for a specific inferencing instance.',
      content: {
        'application/json': {
          schema: z.object({ inference: inferenceInterfaceSchema }),
        },
      },
    },
  },
})

interface GetInferenceService {
  inference: InferenceInterface
}

export const getInference = [
  bodyParser.json(),
  async (req: Request, res: Response<GetInferenceService>) => {
    req.audit = AuditInfo.ViewInference
    const {
      params: { modelId, image, tag },
    } = parse(req, getInferenceSchema)

    const inference = await getInferenceByImage(req.user, modelId, image, tag)

    await audit.onViewInference(req, inference)

    return res.json({
      inference,
    })
  },
]
