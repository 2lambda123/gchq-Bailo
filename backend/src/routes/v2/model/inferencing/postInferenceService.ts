import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { AuditInfo } from '../../../../connectors/v2/audit/Base.js'
import audit from '../../../../connectors/v2/audit/index.js'
import { InferenceInterface } from '../../../../models/v2/Inference.js'
import { createInference } from '../../../../services/v2/inference.js'
import { inferenceInterfaceSchema, registerPath } from '../../../../services/v2/specification.js'
import { parse } from '../../../../utils/v2/validate.js'

export const postInferenceSchema = z.object({
  params: z.object({
    modelId: z.string({
      required_error: 'Must specify model id as param',
    }),
  }),
  body: z.object({
    image: z.string(),
    tag: z.string(),
    description: z.string(),
    settings: z.object({
      processorType: z.string(),
      memory: z.number().optional(),
      port: z.number(),
    }),
  }),
})

registerPath({
  method: 'post',
  path: '/api/v2/model/{modelId}/inference',
  tags: ['inference'],
  description: 'Create a inferencing service within Bailo',
  schema: postInferenceSchema,
  responses: {
    200: {
      description: 'The created inferencing service.',
      content: {
        'application/json': {
          schema: inferenceInterfaceSchema,
        },
      },
    },
  },
})

interface PostInferenceService {
  inference: InferenceInterface
}

export const postInference = [
  bodyParser.json(),
  async (req: Request, res: Response<PostInferenceService>) => {
    req.audit = AuditInfo.CreateInference
    const {
      params: { modelId },
      body,
    } = parse(req, postInferenceSchema)

    const inference = await createInference(req.user, modelId, body)

    await audit.onCreateInference(req, inference)

    return res.json({
      inference,
    })
  },
]
