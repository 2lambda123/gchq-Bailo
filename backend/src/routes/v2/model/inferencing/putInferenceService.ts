import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { InferenceInterface, ProcessorType } from '../../../../models/v2/Inference.js'
import { updateInference } from '../../../../services/v2/inference.js'
import { inferenceInterfaceSchema, registerPath } from '../../../../services/v2/specification.js'
import { parse } from '../../../../utils/v2/validate.js'

export const putInferenceSchema = z.object({
  params: z.object({
    modelId: z.string({
      required_error: 'Must specify model id as param',
    }),
  }),
  body: z.object({
    image: z.string(),
    tag: z.string(),
    description: z.string(),
    settings: z.object({
      processorType: z.nativeEnum(ProcessorType),
      memory: z.number().optional(),
      port: z.number(),
    }),
  }),
})

registerPath({
  method: 'put',
  path: '/api/v2/model/{modelId}/inference',
  tags: ['inference'],
  description: 'Update a inferencing service within Bailo',
  schema: putInferenceSchema,
  responses: {
    200: {
      description: 'The created inferencing service.',
      content: {
        'application/json': {
          schema: inferenceInterfaceSchema,
        },
      },
    },
  },
})

interface PutInferenceService {
  inference: InferenceInterface
}

export const putInference = [
  bodyParser.json(),
  async (req: Request, res: Response<PutInferenceService>) => {
    const {
      params: { modelId },
      body,
    } = parse(req, putInferenceSchema)

    const inference = await updateInference(req.user, modelId, body)
    return res.json({
      inference,
    })
  },
]
