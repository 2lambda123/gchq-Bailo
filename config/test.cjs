module.exports = {
  user: {
    id: 'user',
  },

  schemas: {
    model: 'Minimal Schema v10',
    deployment: 'Minimal Deployment Schema v6',
  },

  samples: {
    binary: '__tests__/example_models/minimal_model/minimal_binary.zip',
    code: '__tests__/example_models/minimal_model/minimal_code.zip',
    uploadMetadata: '__tests__/example_models/minimal_model/minimal_metadata.json',
    deploymentMetadata: '__tests__/example_models/minimal_model/deployment.json',
  },
}
