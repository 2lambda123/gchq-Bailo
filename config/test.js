module.exports = {
  user: {
    id: 'user',
  },

  schemas: {
    model: 'Minimal Schema v10',
    deployment: 'Minimal Deployment Schema v6',
  },
}
