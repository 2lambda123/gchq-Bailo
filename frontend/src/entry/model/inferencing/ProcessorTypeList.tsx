import { Autocomplete, TextField, Typography } from '@mui/material'
import { useGetUiConfig } from 'actions/uiConfig'
import { SyntheticEvent, useMemo } from 'react'
import Loading from 'src/common/Loading'
import MessageAlert from 'src/MessageAlert'

type ProcessorTypeListProps = {
  value: string
  onChange: (value: string) => void
  readOnly?: boolean
}

export default function ProcessorTypeList({ value, onChange, readOnly = false }: ProcessorTypeListProps) {
  const { uiConfig, isUiConfigLoading, isUiConfigError } = useGetUiConfig()

  const processorTypesList = useMemo(
    () => (uiConfig ? ['cpu', ...Object.values(uiConfig.inference.gpus)] : []),
    [uiConfig],
  )
  const readOnlyProcessorTypeList = useMemo(() => {
    return isUiConfigLoading ? (
      <Loading />
    ) : (
      processorTypesList.map((processorType) => <Typography key={processorType}>{processorType}</Typography>)
    )
  }, [processorTypesList, isUiConfigLoading])

  const handleChange = (_event: SyntheticEvent, newValue: string | null) => {
    onChange(newValue || '')
  }

  if (isUiConfigError) {
    return <MessageAlert message={isUiConfigError.info.message} severity='error' />
  }

  return (
    <>
      {readOnly ? (
        readOnlyProcessorTypeList
      ) : (
        <Autocomplete
          loading={isUiConfigLoading}
          data-test='processorTypesAutocomplete'
          options={processorTypesList}
          value={value}
          onChange={handleChange}
          renderInput={(params) => <TextField {...params} size='small' />}
        />
      )}
    </>
  )
}
