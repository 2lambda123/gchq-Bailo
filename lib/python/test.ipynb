{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bailo import Model, Client, Experiment, Schema, SchemaKind\n",
    "\n",
    "# Instantiating the Bailo client\n",
    "\n",
    "client = Client(\"http://127.0.0.1:8080\") # <- INSERT BAILO URL (if not hosting locally)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'id': 'metrics-v5', 'name': 'Metrics v5', 'description': 'Example Description', 'active': True, 'hidden': False, 'kind': 'model', 'jsonSchema': {'$schema': 'http://json-schema.org/draft-07/schema#', 'type': 'object', 'properties': {'overview': {'title': 'Overview', 'description': 'Summary of the model functionality.', 'type': 'object', 'properties': {'modelOverview': {'title': 'What does the model do?', 'description': 'A description of what the model does.', 'type': 'string', 'minLength': 1, 'maxLength': 5000}, 'tags': {'title': 'Descriptive tags for the model.', 'description': 'These tags will be searchable and will help others find this model.', 'type': 'array', 'widget': 'tagSelectorBeta', 'items': {'type': 'string'}, 'uniqueItems': True}}, 'required': [], 'additionalProperties': False}, 'performance': {'title': 'Performance', 'type': 'object', 'properties': {'performanceMetrics': {'title': 'Performance Metrics', 'description': 'List of metrics, values, and the dataset they were evaluated on', 'type': 'array', 'items': {'type': 'object', 'title': '', 'properties': {'dataset': {'title': 'Dataset used', 'type': 'string'}, 'datasetMetrics': {'type': 'array', 'title': 'Dataset Metrics', 'items': {'type': 'object', 'title': '', 'properties': {'name': {'title': 'Metric name', 'description': 'For example: ACCURACY', 'type': 'string'}, 'value': {'title': 'Model performance metric value', 'description': 'For example: 82', 'type': 'number'}}}}}}}}}}, 'required': [], 'additionalProperties': False}, '_id': '65dcc3f1998b9ba92ba1ae0b', 'createdAt': '2024-02-26T17:01:37.900Z', 'updatedAt': '2024-02-26T17:01:37.900Z', '__v': 0}\n"
     ]
    }
   ],
   "source": [
    "json_schema = {\n",
    "    \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
    "    \"type\": \"object\",\n",
    "    \"properties\": {\n",
    "        \"overview\": {\n",
    "            \"title\": \"Overview\",\n",
    "            \"description\": \"Summary of the model functionality.\",\n",
    "            \"type\": \"object\",\n",
    "            \"properties\": {\n",
    "                \"modelOverview\": {\n",
    "                    \"title\": \"What does the model do?\",\n",
    "                    \"description\": \"A description of what the model does.\",\n",
    "                    \"type\": \"string\",\n",
    "                    \"minLength\": 1,\n",
    "                    \"maxLength\": 5000,\n",
    "                },\n",
    "                \"tags\": {\n",
    "                    \"title\": \"Descriptive tags for the model.\",\n",
    "                    \"description\": \"These tags will be searchable and will help others find this model.\",\n",
    "                    \"type\": \"array\",\n",
    "                    \"widget\": \"tagSelectorBeta\",\n",
    "                    \"items\": {\"type\": \"string\"},\n",
    "                    \"uniqueItems\": True,\n",
    "                },\n",
    "            },\n",
    "            \"required\": [],\n",
    "            \"additionalProperties\": False,\n",
    "        },\n",
    "        \"performance\": {\n",
    "            \"title\": \"Performance\",\n",
    "            \"type\": \"object\",\n",
    "            \"properties\": {\n",
    "                \"performanceMetrics\": {\n",
    "                    \"title\": \"Performance Metrics\",\n",
    "                    \"description\": \"List of metrics, values, and the dataset they were evaluated on\",\n",
    "                    \"type\": \"array\",\n",
    "                    \"items\": {\n",
    "                        \"type\": \"object\",\n",
    "                        \"title\": \"\",\n",
    "                        \"properties\": {\n",
    "                            \"dataset\": {\n",
    "                            \"title\": \"Dataset used\",\n",
    "                            \"type\": \"string\"\n",
    "                        },\n",
    "                        \"datasetMetrics\": {\n",
    "                            \"type\": \"array\",\n",
    "                            \"title\": \"Dataset Metrics\",\n",
    "                            \"items\": {\n",
    "                            \"type\": \"object\",\n",
    "                            \"title\": \"\",\n",
    "                            \"properties\": {\n",
    "                                \"name\": {\n",
    "                                \"title\": \"Metric name\",\n",
    "                                \"description\": \"For example: ACCURACY\",\n",
    "                                \"type\": \"string\"\n",
    "                                },\n",
    "                                \"value\": {\n",
    "                                \"title\": \"Model performance metric value\",\n",
    "                                \"description\": \"For example: 82\",\n",
    "                                \"type\": \"number\"\n",
    "                                }\n",
    "                             }\n",
    "                        }\n",
    "                  }\n",
    "                }\n",
    "              }\n",
    "            },\n",
    "            },\n",
    "        },\n",
    "\n",
    "    },\n",
    "    \"required\": [],\n",
    "    \"additionalProperties\": False,\n",
    "}\n",
    "\n",
    "schema = Schema.create(\n",
    "    client=client,\n",
    "    schema_id=\"metrics-v5\",\n",
    "    name=\"Metrics v5\",\n",
    "    description=\"Example Description\",\n",
    "    kind=SchemaKind.MODEL,\n",
    "    json_schema=json_schema\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = Model.create(client=client, name=\"ResNet\", description=\"ResNet\", team_id=\"uncategorised\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
