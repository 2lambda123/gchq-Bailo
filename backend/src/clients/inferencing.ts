import fetch from 'node-fetch'

import config from '../utils/config.js'
import { BadReq, InternalError } from '../utils/error.js'

interface InferenceService {
  modelId: string
  image: string
  tag: string
  port: number
}

export async function createInferenceService(inferenceServiceParams: InferenceService) {
  let res
  try {
    res = await fetch(`${config.inference.connection.host}/api/deploy`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(inferenceServiceParams),
    })
  } catch (err) {
    throw InternalError('Unable to communicate with the inferencing service.', { err })
  }
  if (!res.ok) {
    throw BadReq('Unrecognised response returned by the inferencing service.')
  }
  // TODO - Update return object. For now, we are just checking the status
  return res.json()
}

export async function updateInferenceService(inferenceServiceParams: InferenceService) {
  let res
  try {
    res = await fetch(`${config.inference.connection.host}/api/update`, {
      method: 'PATCH',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(inferenceServiceParams),
    })
  } catch (err) {
    throw InternalError('Unable to communicate with the inferencing service.', { err })
  }

  if (!res.ok) {
    throw BadReq('Unrecognised response returned by the inferencing service.')
  }
  // TODO - Update return object. For now, we are just checking the status
  return res.json()
}
