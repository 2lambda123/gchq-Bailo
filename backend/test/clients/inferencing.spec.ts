import { describe } from 'node:test'
import { test } from 'shelljs'
import { expect, vi } from 'vitest'
import { updateInferenceService, createInferenceService } from '../../src/clients/inferencing.js'

const configMock = vi.hoisted(() => ({
  inference: {
    enabled: true,
    connection: {
      host: 'http://example.com',
    },
  },
}))

vi.mock('../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMock,
}))

vi

describe('clients > inferencing', () => {
  test('createInferencing > success', async () => {

    expect(() => createInferenceService({} as any))
  })

  test('createInferencing > bad response', async () => {
    expect(() => createInferenceService({} as any)).rejects.toThrowError(/^Unrecoginsed response returned by inferencing service./)
  })

   test('createInfererencing > not found', async () => {
    expect(() => createInferenceService({} as any)).rejects.toThrowError(/^Unable to communicate with the inferencing service/)}
  })

  test('updateInferencing > success', async () => {
    expect(() => updateInferenceService({} as any)).rejects.toThrowError(/^Unable to communicate with the inferencing service/)
  })
  test('updateInferencing > bad response', async () => {
    expect(() => updateInferenceService({} as any)).rejects.toThrowError(/^Unable to communicate with the inferencing service/)
  })

  test('updateInferencing > rejected', async () => {
    expect(() => updateInferenceService({} as any)).rejects.toThrowError(/^Unable to communicate with the inferencing service/)
  })
})
