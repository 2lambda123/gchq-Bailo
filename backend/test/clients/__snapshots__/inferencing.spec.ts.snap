// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`clients > inferencing > createInferencing > success 1`] = `
[
  [
    "http://example.com/api/deploy",
    {
      "body": "{}",
      "headers": Headers {
        Symbol(headers list): HeadersList {
          "cookies": null,
          Symbol(headers map): Map {
            "content-type" => {
              "name": "content-type",
              "value": "application/json",
            },
          },
          Symbol(headers map sorted): null,
        },
        Symbol(guard): "none",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`clients > inferencing > updateInferencing > success 1`] = `
[
  [
    "http://example.com/api/update",
    {
      "body": "{}",
      "headers": Headers {
        Symbol(headers list): HeadersList {
          "cookies": null,
          Symbol(headers map): Map {
            "content-type" => {
              "name": "content-type",
              "value": "application/json",
            },
          },
          Symbol(headers map sorted): null,
        },
        Symbol(guard): "none",
      },
      "method": "PATCH",
    },
  ],
]
`;
